### ISSUE TITLE
[NEXT-1128] Client-side cache is not revalidated when calling revalidatePath + redirect in server action

### ISSUE THREAD

**Issue by benrandja-akram** [vercel/next.js#49450]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2
    Binaries:
      Node: 18.16.0
      npm: 9.5.1
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.4.2-canary.0
      eslint-config-next: 13.4.1
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.0.4
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

https://github.com/benrandja-akram/next-js-revalidate-with-redirec-bug

### To Reproduce

- go to profile page
- submit form 
- go to profile page again
- submit form again
- go to profile page 
- Cache is used and profile page shows stale data.  

### Describe the Bug

client side cache is not revalidated When calling revalidatePath with a redirect in server action 

### Expected Behavior

revalidate client side cache when calling revalidatePath + redirect i nserver action

### Which browser are you using? (if relevant)

chrome

### How are you deploying your application? (if relevant)

https://revalidate-next-js-bug.vercel.app/

<sub>[NEXT-1128](https://linear.app/vercel/issue/NEXT-1128/client-side-cache-is-not-revalidated-one-calling-revalidatepath)</sub>

**Comment by arnasofc** [comment]
I have the same issue, any ideas how to fix this cache issue?

**Comment by benrandja-akram** [comment]
@arnasofc a temporary workaound would be calling router.refresh in a client component

**Comment by sannajammeh** [comment]
Ran into this bug as well. This is not just with redirect, its with all revalidatePath usage.

Here's my repro: https://github.com/sannajammeh/app-router-revalidation-not-refreshing

As usual run 
1. `pnpm install`
2. `npx prisma db push`
3. `pnpm dev`
4. Navigate to `/player` in the browser
5. Click "Create Item" button.

Observations:
1. After pressing "Create Item": Root Layout will update with the correct item.
2. Now press "ITEM LINK: <whatever item name>" and type `testing` into input field and press update
3.  Root layout will update again and u see "ITEM LINK: testing" in the left
4. Click on all other links on the left. Every option will return the previous result without update except for `All Items useRouter() 'forceOptimisticNavigation1'`
5. Redo steps 2 - 4 and see that now every single option returns stale data.

Results:
- `Root Layout` - Always updates
- `Child page` - Always stale
- `Child page Layout` - Always stale
- `Child page with router.push` - Always stale
- `Child page with router.push & forceOptimisticNavigation` - Updates once, subsequent updates always stale
- `Server console`: Actually logs prisma performing select query after revalidation.
 
If you continue adding random numbers while updating, you will soon see each link giving you a different number.

This is very weird caching behavior.

**Comment by vasco3** [comment]
it happens to me when I use `Link` but if I use a regular `a` tag then the data renders dynamically (as expected)

**Comment by benrandja-akram** [comment]
@vasco3 using <a href will trigger a hard navigation

**Comment by sannajammeh** [comment]
> @vasco3 using <a href will trigger a hard navigation

That is correct, but we are looking for hard navigation using SPA. That way we wont unmount parent layouts.

**Comment by Fredkiss3** [comment]
@sannajammeh I've tested your reproduction and tested on my own, and what seems to happen is that even if you revalidate multiple pages, only the current page get revalidated : https://github.com/Fredkiss3/test-revalidate-same-page

In my repro, if you are on `/counter` and click `increment` it will revalidate all the data for the current page and for the layout, if you go to `/counter/sub` which revalidates and does a redirect, it does not revalidates any data on the client.

**Comment by sannajammeh** [comment]
> @sannajammeh I've tested your reproduction and tested on my own, and what seems to happen is that even if you revalidate multiple pages, only the current page get revalidated : https://github.com/Fredkiss3/test-revalidate-same-page
> 
> In my repro, if you are on `/counter` and click `increment` it will revalidate all the data for the current page and for the layout, if you go to `/counter/sub` which revalidates and does a redirect, it does not revalidates any data on the client.

You're right. Still doesn't explain the behavior of `forceOptimisticNavigation`, because that actually makes it work in some cases. So the client router javascript obviously has a way to bust the cache, but it choses not to. Only when router push with `forceOptimisticNavigation` it properly removes the cache. 

This looks like a lot of magic with minimal control. I don't see a magical function like revalidatePath doing anything else than setting a response header to signal the paths that should change. This means a native fetch call would be insufficient in telling Next to bust the cache. I have no idea if they have rolled their own fetch api on the client as well which consumes these headers, but if not, there is no way for Next to actually know what to revalidate.

Problem 2 arises as a result. We have no way to explicitly telling Next to bust cache on the client. There should be a `router.revalidatePath() & router.revalidateTag()` api on the client side router, not a one size fits all `forceOptimisticNavigation` that only works on router push.

On a second pass, here are the response headers from a POST using server actions and 3 revalidatePath calls:

```
cache-control: no-store, must-revalidate
connection: close
content-type: text/x-component
date: Sat, 13 May 2023 00:01:07 GMT
keep-alive: timeout=5
transfer-encoding: chunked
vary: RSC, Next-Router-State-Tree, Next-Router-Prefetch, Accept-Encoding
x-powered-by: Next.js
```

From this, the only thing I can see Next handling is `cache-control` header. Meaning it really has no idea what to update on the frontend except for the requesting path.

**Comment by benrandja-akram** [comment]
[It's an internal API, hence it's undocumented. It's not meant to be used by the end user.
](https://github.com/vercel/next.js/issues/47894#issuecomment-1495856463)
@sannajammeh 
by looking to this comment, `forceOptimisticNavigation` seems to be an internal API not meant to be used in user land.

**Comment by Fredkiss3** [comment]
@sannajammeh You have to inspect not the headers but the body, the return of an action which revalidates data is the same format as the payload for server components which are also used for navigations btw. something like this :

```
0:[["children","__PAGE__",["__PAGE__",{}],"$L1",[[],["$L2",["$","meta",null,{"name":"next-size-adjust"}]]]]]
1:["$L3",null]
2:[[["$","meta",null,{"charSet":"utf-8"}],["$","title",null,{"children":"The best todo App in the world"}],null,null,null,null,null,null,null,null,null,["$","meta",null,{"name":"viewport","content":"width=device-width, initial-scale=1"}],null,null,null,null,null,null,null,null,null,null,[]],[null,null,null,null],null,null,[null,null,null,null,null],null,null,null,null,null]
3:["$","main",null,{"className":"p-8 flex flex-col items-center","children":["$","div",null,{"className":"flex flex-col items-stretch gap-4 max-w-[400px] w-full","children":[["$","h1",null,{"className":"text-4xl font-bold ","children":"TODO APP "}],"$L4"]}]}]
5:I{"id":"(app-client)/./src/app/todo-app-client.tsx","chunks":["app/page:static/chunks/app/page.js"],"name":"TodoAppClient","async":false}
4:["$","$L5",null,{"todos":[{"id":"ff085847-1e6e-45fe-9acb-1a1e011c16c8","label":"Aller chez Hilary","dueDate":"2023-05-08T00:00:00.000Z","completed":false},{"id":"fd09f38e-76dd-4b58-adb3-f2048c4a2050","label":"go to abidjan","dueDate":"2023-05-09T00:00:00.000Z","completed":true},{"id":"7cec5500-1f32-4f17-acb5-ce431d706d59","label":"code","dueDate":"2023-05-09T00:00:00.000Z","completed":true}]}]
```

What i think next does : 

1. client calls the server with `POST`
2. server calls `revalidatePath` (or redirect) which marks that current path needs to be revalidated
3. server revalidates data and produces the rendered payload for the client
4. server send the rendered payload to the client
5. client rerender with new data and bust the cache for the current page

this is more or less confirmed by the code in this PR https://github.com/vercel/next.js/pull/49582/files :

```ts
 // ... code above
 const staticGenerationAsyncStore = (fetch as any)
    .__nextGetStaticStore?.()
    ?.getStore() as undefined | StaticGenerationStore
  if (staticGenerationAsyncStore) {
    staticGenerationAsyncStore.pathWasRevalidated = true // <--- THIS CODE HERE
  }
 // ... code below
```

**Comment by sannajammeh** [comment]
@Fredkiss3 

Thanks for clarifying. I see you're right. I assume step 5. is where this all falls apart.

**Comment by Fredkiss3** [comment]
It does work as intended if you only call `revalidatePath`, but it does fall appart if used with `redirect`, and there is also the second problem of only revalidating the cache for 1 page, even if we call `revalidatePath` multiple times.

**Comment by sannajammeh** [comment]
> It does work as intended if you only call `revalidatePath`, but it does fall appart if used with `redirect`, and there is also the second problem of only revalidating the cache for 1 page, even if we call `revalidatePath` multiple times.

From what I've seen in my repro + others is that it's not just for 1 page. It only works on the **current** path.  So even a single `revalidatePath` call is not working as intended. 

Example:
Current url `/dashboard/posts/new`
revalidatePath called on `/dashboard/posts`
Next Link navigation to `/dashboard/posts` === stale data.

This cannot possibly be intended.

**Comment by Fredkiss3** [comment]
Yeah, you are totally right, only for the current path, i also edited my comment with some observations.

**Comment by eric-burel** [comment]
> @vasco3 using <a href will trigger a hard navigation

I think the difference with a raw anchor (or `window.location.href`) and a hard refresh with `router.push` is that with `router.push` it may only refetch the relevant segments. So if you have a route `/[lang]/[fooId]/[barId]`, you navigate "/fr/129/341" and update the "bar" of id "341", you want only the data for the most nested layout to be loaded. Which doesn't seem feasible right now. Do I understand the issue correctly?

**Comment by vasco3** [comment]
@eric-burel yeah I want. to revalidate nested layout when I use a server action when submitting a form. The client state isn't updated with the revalidatePath so even if I navigate (with Link) back and forth of different routes the data is stale. The only way to get fresh data is to navigate with `a` anchor or I assume `window.location.href`.

**Comment by sannajammeh** [comment]
This may or may not be related, but it appears the client cache on soft navigation is stale regardless. 

After performing a `revalidatePath("/url")`

With default Next cache (aka no custom config) options the following happens :

- Direct request of `/url`: works after revalidation as expected
- Hard navigation: Works after revalidation as expected
- Direct request of `/url`, followed by soft navigation to other page and soft navigation back to `/url`: Stale data
- Direct request of INDEX, followed by soft navigation to `/url`: Stale data
- Setting the entire `/url` to `fetchCache = "force-no-store"` fixes this issue.

This only happens when using the fetch API it appears. I'll see if I can make a repro tomorrow.

**Comment by riyaadh-abrahams** [comment]
> > It does work as intended if you only call `revalidatePath`, but it does fall appart if used with `redirect`, and there is also the second problem of only revalidating the cache for 1 page, even if we call `revalidatePath` multiple times.
> 
> From what I've seen in my repro + others is that it's not just for 1 page. It only works on the **current** path. So even a single `revalidatePath` call is not working as intended.
> 
> Example: Current url `/dashboard/posts/new` revalidatePath called on `/dashboard/posts` Next Link navigation to `/dashboard/posts` === stale data.

for some reason. when I replace revalidatePath with redirect inside the server action, thinks seem to work. (But does not redirect)

Example:
Current url /dashboard/projects/new
`redirect` called on /dashboard/projects
Next Link navigation to /dashboard/posts === ✅correct data.

**Comment by Jordaneisenburger** [comment]
> This may or may not be related, but it appears the client cache on soft navigation is stale regardless.
> 
> After performing a `revalidatePath("/url")`
> 
> With default Next cache (aka no custom config) options the following happens :
> 
> * Client side revalidation of `/url` as a result of `revalidatePath`: works as expected ✅
> * Direct request of `/url`: works after revalidation as expected ✅
> * Hard navigation: Works after revalidation as expected ✅
> * Direct request of `/url`, followed by soft navigation to other page and soft navigation back to `/url`: Stale data ❌
> * Direct request of INDEX, followed by soft navigation to `/url`: Stale data ❌
> * Soft navigation to `/` then soft navigation back to `/url`: Stale data ❌
> * Setting the entire `/url` to `fetchCache = "force-no-store"` fixes this issue. ✅
> 
> This only happens when using the fetch API it appears. I'll see if I can make a repro tomorrow. It appears soft navigation is the culprit here.

Experiencing the exact same issue, but I'm not sure if you have a `/account/users` page that shows some user data and you then navigate to `/account/users/[id]`  where you edit this data in a Server Action and call `revalidatePath('/account/users');`  and then use a <Link /> to navigate back to `/account/users` it's supposed to be updated data? 

I think the "good to know" section explains that this is indeed not the case for static routes like "/account/users" it only works that way for dynamic segments and all pages after/below the dynamic segment see https://nextjs.org/docs/app/api-reference/functions/revalidatePath

**Comment by shuding** [comment]
This was fixed by #50848 in the latest canary. Let us know if it's still not working correctly, thanks!
